code/modules/projectiles/gun.dm

```diff
@@ -249,52 +249,65 @@
 	update_icon()
 	return ..()
 
-/obj/item/gun/proc/Fire(atom/target, mob/living/user, clickparams, pointblank=0, reflex=0, set_click_cooldown=TRUE)
-	if(!user || !target)
+/obj/item/gun/proc/Fire(atom/target, atom/movable/firer, clickparams, pointblank = FALSE, reflex = FALSE, set_click_cooldown = TRUE, target_zone = BP_CHEST)
+	if(!firer || !target)
 		return
-	if(target.z != user.z)
+	if(target.z != firer.z)
 		return
 
-	add_fingerprint(user)
-
-	if((!waterproof && submerged()) || !special_check(user))
+	if((!waterproof && submerged()))
 		return
 
-	if(safety())
-		if(user.a_intent == I_HURT && !user.skill_fail_prob(SKILL_WEAPONS, 100, SKILL_EXPERT, 0.5)) //reflex un-safeying
-			toggle_safety(user)
-		else
-			handle_click_empty(user)
-			return
-
 	if(world.time < next_fire_time)
 		if (world.time % 3) //to prevent spam
-			to_chat(user, "<span class='warning'>[src] is not ready to fire again!</span>")
+			to_chat(firer, "<span class='warning'>[src] is not ready to fire again!</span>")
 		return
 
-	last_safety_check = world.time
+	if(isliving(firer))
+		var/mob/living/user = firer
+		add_fingerprint(user)
+
+		if(!special_check(user))
+			return
+
+		if(safety())
+			if(user.a_intent == I_HURT && !user.skill_fail_prob(SKILL_WEAPONS, 100, SKILL_EXPERT, 0.5)) //reflex un-safeying
+				toggle_safety(user)
+			else
+				handle_click_empty(user)
+				return
+
+		last_safety_check = world.time
+	
 	if(set_click_cooldown)
 		var/shoot_time = (burst - 1) * burst_delay
-		user.setClickCooldown(shoot_time) //no clicking on things while shooting
+		if(isliving(firer))
+			var/mob/living/user = firer
```

Comment:
Might be worth declaring `user` up the top somewhere, then just doing `istype(user)` in the spots where you're doing `isliving(firer)`.